lib/parsers/json_parser/validator.rb

```diff
@@ -0,0 +1,318 @@
+    accepted = next_step.call(c)
+    reject_char(c) unless accepted
+  end
+
+  def read_whitespace(c)
+    whitespace?(c)
+  end
+
+  def read_colon(c)
+    if c == ":"
+      @current_state = :awaiting_object_attribute_value
+      return true
+    end
+    false
+  end
+
+  def read_valid_string_char(c)
+    if @escape_next
+      @escape_next = false
+      return true
+    end
+
+    if c == ESCAPE_CHAR
+      @escape_next = true
+      return true
+    end
+    !control_char?(c) and c != "\""
+  end
+
+  def read_valid_literal_char(c)
+    if valid_literal_char?(c)
+      @current_literal_size += 1
+      return true
+    end
+
+    false
+  end
+
+  def read_comma_separator(c)
+    if c == ","
+      @current_state = :awaiting_object_attribute_key if @current_node == :object
+      @current_state = :awaiting_array_value if @current_node == :array
+      return true
+    end
+    false
+  end
+
+  # Object: {"k1":"val", "k2":[1,2,3], "k4": undefined, "k5": {"l1": 6}}
+  def start_object(c)
+    return false if whitespace?(c)
+    return false unless c == "{"
+
+    begin_node(:object)
+    @current_state = :awaiting_object_attribute_key
+    true
+  end
+
+  def close_object(c)
+    return false if whitespace?(c)
+    return false unless @current_node == :object and c == "}"
+
+    end_node
+    @current_state = :awaiting_next_or_close unless @current_node.nil?
+    true
+  end
+
+  # Array: [1, "two", true, undefined, {}, []]
+  def start_array(c)
+    return false unless c == "["
+
+    begin_node(:array)
+    @current_state = :awaiting_array_value
+    true
+  end
+
+  def close_array(c)
+    return false if whitespace?(c)
+    return false unless @current_node == :array and c == "]"
+
+    end_node
+    @current_state = :awaiting_next_or_close unless @current_node.nil?
+    true
+  end
+
+  def start_attribute_key(c)
+    return false unless c == "\""
+
+    begin_node(:string)
+    @current_state = :reading_object_attribute_key
+    true
+  end
+
+  def close_attribute_key(c)
+    return false if @escape_next
+    return false unless c == "\""
+    end_node
+    @current_state = :awaiting_object_colon_separator
+    true
+  end
+
+  # Strings: "Foo"
+  def start_string(c)
+    return false unless c == "\""
+
+    begin_node(:string)
+    @current_state = :reading_string
+    true
+  end
+
+  def close_string(c)
+    return false if @escape_next
+    return false unless c == "\""
+    end_node
+    @current_state = :awaiting_next_or_close
+    true
+  end
+
+  # literals: null, undefined, true, false, NaN, infinity, -123.456e10 -123,456e10
+  def start_literal(c)
+    return false unless valid_literal_char?(c)
+
+    begin_node(:literal)
+    @current_state = :reading_literal
+    @current_literal_size = 1
+    true
+  end
+
+  def close_literal(c)
+    raise JSONParserError, "Literal to large at #{@pos}" if @current_literal_size > MAX_LITERAL_SIZE
+
+    if whitespace?(c) or c == "," or c == "]" or c == "}"
```

Comment:
Usually `||` is preferred, unless there a specific reason to use `or`. I'd be inclined to make the same suggestion for the `when_its`s above (is it like just for readability or it's an actual precedence requirement?)

Reply:
It was a deliberate decision to use `or` with `when_its` for the sake of looking like english statements. does that one look weird?

here, I'll replace this with an array `include?` instead, actually... but I agree `||` is better than `or` for the use case.

Reply:
If we're sure the resulting precedence works in our favour, `or`s are just fine 😊